File: flutter_app/lib/services/logger.dart
import 'dart:developer' as dev;
import 'dart:convert';

class LogEntry {
  final DateTime timestamp;
  final String level;
  final String source;
  final String message;
  final Map<String, dynamic>? data;

  LogEntry({
    required this.timestamp,
    required this.level,
    required this.source,
    required this.message,
    this.data,
  });

  Map<String, dynamic> toJson() => {
    't': timestamp.toIso8601String(),
    'l': level,
    's': source,
    'm': message,
    'd': data ?? {},
  };
}

class Logger {
  final String source;
  final List<LogEntry> _buffer = [];
  static const int _maxEntries = 500;
  static final List<void Function(LogEntry)> _listeners = [];

  Logger(this.source);

  static void addListener(void Function(LogEntry) cb) {
    _listeners.add(cb);
  }

  void _log(String level, String message, [Map<String, dynamic>? data]) {
    final entry = LogEntry(
      timestamp: DateTime.now(),
      level: level,
      source: source,
      message: message,
      data: data,
    );
    _buffer.add(entry);
    if (_buffer.length > _maxEntries) _buffer.removeAt(0);
    for (final cb in _listeners) {
      try { cb(entry); } catch (_) {}
    }
    dev.log('[${source}] $message', name: 'FCSTN');
  }

  void info(String message, [Map<String, dynamic>? data]) => _log('INFO', message, data);
  void warn(String message, [Map<String, dynamic>? data]) => _log('WARN', message, data);
  void error(String message, [Map<String, dynamic>? data]) => _log('ERROR', message, data);
  void debug(String message, [Map<String, dynamic>? data]) => _log('DEBUG', message, data);

  List<LogEntry> recent([int n = 100]) {
    return _buffer.length > n ? _buffer.sublist(_buffer.length - n) : _buffer;
  }
}

File: shared/logger.py
import json, time, logging, threading, os
from collections import deque
from dataclasses import dataclass, field, asdict
from typing import Optional

@dataclass
class LogEntry:
    timestamp: float
    level: str
    source: str
    message: str
    data: dict = field(default_factory=dict)

class RingLogger:
    def __init__(self, capacity=2000):
        self._entries = deque(maxlen=capacity)
        self._lock = threading.Lock()
        self._listeners = []

    def log(self, level, source, message, data=None):
        entry = LogEntry(
            timestamp=time.time(),
            level=level,
            source=source,
            message=message,
            data=data or {},
        )
        with self._lock:
            self._entries.append(entry)
        for cb in self._listeners:
            try:
                cb(entry)
            except Exception:
                pass

    def info(self, source, message, data=None):
        self.log("INFO", source, message, data)

    def warn(self, source, message, data=None):
        self.log("WARN", source, message, data)

    def error(self, source, message, data=None):
        self.log("ERROR", source, message, data)

    def debug(self, source, message, data=None):
        self.log("DEBUG", source, message, data)

    def recent(self, n=100, level=None):
        with self._lock:
            entries = list(self._entries)
        if level:
            entries = [e for e in entries if e.level == level]
        return entries[-n:]

    def to_dict_list(self, n=100):
        return [asdict(e) for e in self.recent(n)]

    def on_log(self, callback):
        self._listeners.append(callback)

    def remove_listener(self, callback):
        if callback in self._listeners:
            self._listeners.remove(callback)

ROOT_LOG = RingLogger(capacity=2000)

def get_logger(source):
    class _Logger:
        def info(self, msg, data=None):
            ROOT_LOG.info(source, msg, data)
            logging.info(f"[{source}] {msg}")
        def warn(self, msg, data=None):
            ROOT_LOG.warn(source, msg, data)
            logging.warning(f"[{source}] {msg}")
        def error(self, msg, data=None):
            ROOT_LOG.error(source, msg, data)
            logging.error(f"[{source}] {msg}")
        def debug(self, msg, data=None):
            ROOT_LOG.debug(source, msg, data)
            logging.debug(f"[{source}] {msg}")
    return _Logger()
